# HG changeset patch
# User Rik <rik@octave.org>
# Date 1550248196 28800
#      Fri Feb 15 08:29:56 2019 -0800
# Node ID f85d491d232270ccb5fad8cf015d7a1bb5050e7f
# Parent  75d79c39ac9226933067a4eb56317df092c1280b
# Parent  da8ebed6d56e242a173f270eac9963bf4a2bb364
maint: merge stable to default.

diff --git a/libgui/src/documentation.cc b/libgui/src/documentation.cc
--- a/libgui/src/documentation.cc
+++ b/libgui/src/documentation.cc
@@ -72,16 +72,21 @@ namespace octave
       collection = QString::fromStdString (config::oct_doc_dir ()
                                            + sys::file_ops::dir_sep_str ()
                                            + "octave_interpreter.qhc");
 
     // Setup the help engine with the original collection, use a writable copy
     // of the original collection and load the help data
     m_help_engine = new QHelpEngine (collection, this);
 
+    // Mark help as readonly to avoid error if collection file is stored in a
+    // readonly location
+    m_help_engine->setProperty ("_q_readonly",
+                                QVariant::fromValue<bool> (true));
+
     QString tmpdir = QDir::tempPath();
     m_collection
       = QString::fromStdString (sys::tempnam (tmpdir.toStdString (),
                                               "oct-qhelp-"));
 
     if (m_help_engine->copyCollectionFile (m_collection))
       m_help_engine->setCollectionFile (m_collection);
     else
diff --git a/libinterp/corefcn/svd.cc b/libinterp/corefcn/svd.cc
--- a/libinterp/corefcn/svd.cc
+++ b/libinterp/corefcn/svd.cc
@@ -154,20 +154,20 @@ decomposition, eliminating the unnecessa
 
 If the second argument is exactly 0, then the choice of decomposition is based
 on the matrix @var{A}.  If @var{A} has more rows than columns then an
 economy-sized decomposition is returned, otherwise a regular decomposition
 is calculated.
 
 Algorithm Notes: When calculating the full decomposition (left and right
 singular matrices in addition to singular values) there is a choice of two
-routines in @sc{lapack}.  The default routine used by Octave is @code{gesdd}
-which is 5X faster than the alternative @code{gesvd}, but may use more memory
-and may be less accurate for some matrices.  See the documentation for
-@code{svd_driver} for more information.
+routines in @sc{lapack}.  The default routine used by Octave is @code{gesvd}.
+The alternative is @code{gesdd} which is 5X faster, but may use more memory
+and may be inaccurate for some input matrices.  See the documentation for
+@code{svd_driver} for more information on choosing a driver.
 @seealso{svd_driver, svds, eig, lu, chol, hess, qr, qz}
 @end deftypefn */)
 {
   int nargin = args.length ();
 
   if (nargin < 1 || nargin > 2 || nargout > 3)
     print_usage ();
 
diff --git a/scripts/plot/draw/colorbar.m b/scripts/plot/draw/colorbar.m
--- a/scripts/plot/draw/colorbar.m
+++ b/scripts/plot/draw/colorbar.m
@@ -438,17 +438,17 @@ function cb_restore_axes (hcb, ~, hax, o
     set (hax, "__colorbar_handle__", []);
   endif
 
 endfunction
 
 ## Update colorbar when clim has changed
 function cb_clim (hax, ~, hcb, hi)
 
-  if (isaxes (hax))
+  if (isaxes (hax) && isaxes (hcb))
     clen = rows (get (hax, "colormap"));
     cext = get (hax, "clim");
     cdiff = (cext(2) - cext(1)) / clen / 2;
     cmin = cext(1) + cdiff;
     cmax = cext(2) - cdiff;
 
     if (strcmp (get (hcb, "__vertical__"), "on"))
       set (hi, "ydata", [cmin, cmax]);
diff --git a/scripts/plot/util/private/__gnuplot_get_var__.m b/scripts/plot/util/private/__gnuplot_get_var__.m
--- a/scripts/plot/util/private/__gnuplot_get_var__.m
+++ b/scripts/plot/util/private/__gnuplot_get_var__.m
@@ -100,32 +100,35 @@ function gp_var_value = __gnuplot_get_va
         endif
       endwhile
       str = strcat (str{:});
       fclose (gpin);
     else
       ## Direct gnuplot to print to <STDOUT>
       fprintf (ostream, ['set print "-";' "\n"]);
       fflush (ostream);
-      gp_cmd = sprintf (["\n" 'if (exists("%s")) print "OCTAVE: ", %s; else print NaN' "\n"],
+      gp_cmd = sprintf (["\n" 'if (exists("%s")) print "OCTAVE: ", %s, ' ...
+                        '" :END_OCTAVE"; else print NaN' "\n"],
                         gp_var_name(1:n), gp_var_name);
       fputs (ostream, gp_cmd);
       fflush (ostream);
       ## Direct gnuplot to print to <STDERR>
       fputs (ostream, "set print;\n");
       fflush (ostream);
 
-      str = {};
-      while (isempty (str))
-        str = fread (istream, "*char")';
-        if (isempty (str))
-          pause (0.05);
-        else
-          str = regexp (str, 'OCTAVE:.*', "match");
-          str = str{end}(8:end);
+      str = "";
+      t_start = tic ();
+      while (toc (t_start) < 10)
+        str = [str, fread(istream, "*char")'];
+        if (! isempty (str))
+          re_str = regexp (str, "OCTAVE: (.*) :END_OCTAVE", "tokens");
+          if (! isempty (re_str))
+            str = re_str{end}{1};
+            break;
+          endif
         endif
         fclear (istream);
       endwhile
     endif
 
     ## Strip out EOLs and the continuation character "|"
     str(str=="\n" | str=="\r") = "";
     n_continue = strfind (str, " \\ ");
